# Benchmark Cognitive Services Azure (Text Analytics)



In [1]:
from azure.ai.textanalytics import TextAnalyticsClient
from azure.core.credentials import AzureKeyCredential
import pandas as pd
import os
import joblib
from sklearn.linear_model import LogisticRegression

In [2]:
key = os.environ["COGNITIVE_AZURE_KEY"]
endpoint = os.environ["COGNITIVE_AZURE_ENDPOINT"]

In [3]:
def authenticate_client():
    ta_credential = AzureKeyCredential(key)
    text_analytics_client = TextAnalyticsClient(
            endpoint=endpoint, 
            credential=ta_credential)
    return text_analytics_client

client = authenticate_client()

In [4]:
def sentiment_analysis_example(client):

    documents = ["I had the best day of my life. I wish you were there with me."]
    response = client.analyze_sentiment(documents=documents)[0]
    print("Document Sentiment: {}".format(response.sentiment))
    print("Overall scores: positive={0:.2f}; neutral={1:.2f}; negative={2:.2f} \n".format(
        response.confidence_scores.positive,
        response.confidence_scores.neutral,
        response.confidence_scores.negative,
    ))
    for idx, sentence in enumerate(response.sentences):
        print("Sentence: {}".format(sentence.text))
        print("Sentence {} sentiment: {}".format(idx+1, sentence.sentiment))
        print("Sentence score:\nPositive={0:.2f}\nNeutral={1:.2f}\nNegative={2:.2f}\n".format(
            sentence.confidence_scores.positive,
            sentence.confidence_scores.neutral,
            sentence.confidence_scores.negative,
        ))
          
sentiment_analysis_example(client)

Document Sentiment: positive
Overall scores: positive=1.00; neutral=0.00; negative=0.00 

Sentence: I had the best day of my life.
Sentence 1 sentiment: positive
Sentence score:
Positive=1.00
Neutral=0.00
Negative=0.00

Sentence: I wish you were there with me.
Sentence 2 sentiment: neutral
Sentence score:
Positive=0.21
Neutral=0.77
Negative=0.02



In [5]:
def send_request(document):
    if len(document) > 10:
        all_doc_send=[]
        response=[]
        range_10 = [i for i in range(0,len(document)+1,10)]
        for i in range(1,len(range_10)):
            all_doc_send.append(document[range_10[i - 1 ]:range_10[i]])
        for doc in all_doc_send:
            response.extend(client.analyze_sentiment(documents=doc))
    else:
        response = client.analyze_sentiment(documents=document)
    return response

def format_response(response):
    for resp in response:
        print("Document Sentiment: {}".format(resp.sentiment))
        print("Overall scores: positive={0:.2f}; neutral={1:.2f}; negative={2:.2f} \n".format(
            resp.confidence_scores.positive,
            resp.confidence_scores.neutral,
            resp.confidence_scores.negative,
        ))
        
def get_pred(resp,get_neutral=False):
    positive = resp.confidence_scores.positive
    neutral = resp.confidence_scores.neutral
    negative = resp.confidence_scores.negative
    if get_neutral :
        if (positive > neutral) & (positive > negative):
            return "good"
        elif (negative > neutral) & (negative > positive):
            return "bad"
        else:
            return "neutral"
    else:
        if positive > negative:
            return "good"
        else:
            return "bad"
    
def get_response_prediction(response,get_neutral=False):
    dic={
        "good_score":[],
        "neutral_score":[],
        "bad_score":[],
        "prediction":[],
    }
    for resp in response:
        dic["good_score"].append(resp.confidence_scores.positive)
        dic["neutral_score"].append(resp.confidence_scores.neutral)
        dic["bad_score"].append(resp.confidence_scores.negative)
        dic["prediction"].append(get_pred(resp,get_neutral=get_neutral))
    return dic

# Benchmark azure cognitive

In [6]:
test_sample = pd.read_csv("./data/sample_test_dataset.csv",index_col=0)
print("Dimensionnalité :",test_sample.shape)
test_sample.head()

Dimensionnalité : (4000, 2)


text sentiment
0  Chillin @ the beach with my girl brit-brit wat...      good
1  @indraherlambang really wanna say that ure lik...      good
2  @ambermac Gotta love Brad Pitt's performance, ...      good
3  Just realized that Matthew Sweet &amp; Susanna...      good
4  @marksphone sorry, meant that to come from my ...      good

In [7]:
test_sample["sentiment"].value_counts()

good    2000
bad     2000
Name: sentiment, dtype: int64

## tous les tweets

Le coût de ce service est de 0.87€ pour 1000 requêtes, avec ce test nous somme à 4000 requêtes sois un total de 3.48€

In [8]:
document_send = test_sample["text"].tolist()
y_true = test_sample["sentiment"].tolist()

In [9]:
# dic_response = get_response_prediction(send_request(document_send))
# joblib.dump(dic_response,"./data/cognitive_service_response.pkl")


In [15]:
dic_response = joblib.load("./data/cognitive_service_response.pkl")

['./data/cognitive_service_response.pkl']

In [13]:
good_score = dic_response["good_score"]
neutral_score = dic_response["neutral_score"]
bad_score = dic_response["bad_score"]

In [30]:
import numpy as np
lr = LogisticRegression()
feature_cogn = np.array([good_score,neutral_score,bad_score]).T
lr.fit(feature_cogn,y_true)

In [ ]:
y_pred = lr.predict(feature_cogn)
y_pred_proba = lr.predict_proba(feature_cogn)

In [36]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, roc_auc_score

def show_result_binary(y_pred, y_pred_prob, y_true):
    print("accuracy score :",accuracy_score(y_true,y_pred))
    print("AUC score :",roc_auc_score(y_true,y_pred_prob[:,1]))
    cm = confusion_matrix(y_true,y_pred)
    print(classification_report(y_true,y_pred))
    return pd.DataFrame(cm,columns=["good","bad"],index=["good","bad"])

In [37]:
show_result_binary(y_pred,y_pred_proba,y_true)

accuracy score : 0.71475
AUC score : 0.7786213750000001
              precision    recall  f1-score   support

         bad       0.77      0.61      0.68      2000
        good       0.68      0.82      0.74      2000

    accuracy                           0.71      4000
   macro avg       0.73      0.71      0.71      4000
weighted avg       0.73      0.71      0.71      4000



good   bad
good  1216   784
bad    357  1643

Voici les résultats du service cognitif d'azure, on vois ici que les résultats sont globalement assez bon notament pour la détection des bad tweet avec une précision de 0.77 des bad tweet le score AUC est ici de 0.7786 qui est un score correct mais pas très performant.
Pour résumer le service cognitif d'azure est un outil très généraliste et permet d'avoir un service de détéction des sentiment rapidement et globalement assez efficace, cependant, les résultats sont loin d'être optimaux et pour des frais d'utilisation assez consequent pour une utilisation soutenue du service.